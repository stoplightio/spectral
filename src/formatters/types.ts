import { IRuleResult } from '../types';

export type Formatter = (results: IRuleResult[]) => string;
