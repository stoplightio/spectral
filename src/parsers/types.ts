import { GetLocationForJsonPath, IParserResult } from '@stoplight/types';

export interface IParser<R extends IParserResult = IParserResult<unknown>> {
  parse(input: string): R;
  getLocationForJsonPath: GetLocationForJsonPath<R>;
  trapAccess<T extends object = object>(obj: T): T;
}
