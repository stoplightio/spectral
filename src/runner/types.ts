import type { DocumentInventory } from '../documentInventory';
import type { Ruleset } from '../ruleset/ruleset';
import { ISpectralDiagnostic } from '../types';

export interface IRunnerInternalContext {
  ruleset: Ruleset;
  documentInventory: DocumentInventory;
  results: ISpectralDiagnostic[];
  promises: Array<Promise<void>>;
}
