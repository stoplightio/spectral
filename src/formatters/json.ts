import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
