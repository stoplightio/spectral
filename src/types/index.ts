export * from './spectral';
export * from './function';
