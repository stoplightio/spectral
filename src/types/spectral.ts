import { IResolveOpts, IResolveResult } from '@stoplight/json-ref-resolver/types';
import { DiagnosticSeverity, Dictionary, IDiagnostic, JsonPath } from '@stoplight/types';
import { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';
import { IFunction, IProcessedRule, IRule } from '.';
import { Rule } from '../rule';
import { ComputeFingerprintFunc } from '../utils';

export type FunctionCollection = Dictionary<IFunction<any>, string>;
export type RuleCollection = Dictionary<IProcessedRule, string>;
export type PartialRuleCollection = Dictionary<Partial<IRule>, string>;
export type RunRuleCollection = Dictionary<Rule, string>;

export type SpectralDiagnosticSeverity = DiagnosticSeverity | -1;

/**
 * Name of the rule with a boolean value to enable or disable the rule.
 *
 * Will expand on this format later to allow for things like overriding rule options.
 */
export type RuleDeclarationCollection = Dictionary<boolean, string>;

export interface IConstructorOpts {
  resolver?: IResolver;
  computeFingerprint?: ComputeFingerprintFunc;
  useNimma?: boolean;
  proxyUri?: string;
}

export interface IRunOpts {
  ignoreUnknownFormat?: boolean;
  resolve?: {
    documentUri?: string;
  };
}

export interface IRuleResult extends IDiagnostic {
  path: JsonPath;
}

export interface ISpectralFullResult {
  resolved: unknown;
  results: IRuleResult[];
}

export interface IGivenNode {
  path: JsonPath;
  value: any;
}

export type ResolveResult = Omit<IResolveResult, 'runner'>;

export interface IResolver {
  resolve(source: unknown, opts?: IResolveOpts): Promise<ResolveResult>;
}

// todo(@p0lip): make it string | null when working on 6.0
export type FormatLookup = (document: unknown, source?: string) => boolean;
export type RegisteredFormats = Dictionary<FormatLookup, string>;

export type JSONSchema = JSONSchema4 | JSONSchema6 | JSONSchema7;
