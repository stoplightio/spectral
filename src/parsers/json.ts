import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
