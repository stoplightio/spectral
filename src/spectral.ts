import { safeStringify } from '@stoplight/json';
import { Resolver } from '@stoplight/json-ref-resolver';
import { DiagnosticSeverity, Dictionary } from '@stoplight/types';
import { YamlParserResult } from '@stoplight/yaml';
import { memoize, merge } from 'lodash';

import { STATIC_ASSETS } from './assets';
import { Document, IDocument, IParsedResult, isParsedResult, ParsedDocument } from './document';
import { DocumentInventory } from './documentInventory';
import { functions as defaultFunctions } from './functions';
import * as Parsers from './parsers';
import { readRuleset } from './rulesets';
import { compileExportedFunction } from './rulesets/evaluators';
import { mergeExceptions } from './rulesets/mergers/exceptions';
import { IRulesetReadOptions } from './rulesets/reader';
import { DEFAULT_SEVERITY_LEVEL, getDiagnosticSeverity } from './rulesets/severity';
import { runRules } from './runner';
import {
  FormatLookup,
  FunctionCollection,
  IConstructorOpts,
  IResolver,
  IRuleResult,
  IRunOpts,
  ISpectralFullResult,
  PartialRuleCollection,
  RegisteredFormats,
  RuleCollection,
  RunRuleCollection,
} from './types';
import { IRuleset, RulesetExceptionCollection } from './types/ruleset';
import { ComputeFingerprintFunc, defaultComputeResultFingerprint, empty, prepareResults } from './utils';
import { generateDocumentWideResult } from './utils/generateDocumentWideResult';

memoize.Cache = WeakMap;

export * from './types';

export class Spectral {
  private readonly _resolver: IResolver;

  public functions: FunctionCollection = { ...defaultFunctions };
  public rules: RunRuleCollection = {};
  public exceptions: RulesetExceptionCollection = {};
  public formats: RegisteredFormats;

  private readonly _computeFingerprint: ComputeFingerprintFunc;

  constructor(opts?: IConstructorOpts) {
    this._computeFingerprint = memoize(opts?.computeFingerprint || defaultComputeResultFingerprint);
    this._resolver = opts?.resolver || new Resolver();
    this.formats = {};
  }

  public static registerStaticAssets(assets: Dictionary<string, string>) {
    empty(STATIC_ASSETS);
    Object.assign(STATIC_ASSETS, assets);
  }

  public async runWithResolved(
    target: IParsedResult | IDocument | object | string,
    opts: IRunOpts = {},
  ): Promise<ISpectralFullResult> {
    const document: IDocument =
      target instanceof Document
        ? target
        : isParsedResult(target)
        ? new ParsedDocument(target)
        : new Document<unknown, YamlParserResult<unknown>>(
            typeof target === 'string' ? target : safeStringify(target, undefined, 2),
            Parsers.Yaml,
            opts.resolve?.documentUri,
          );

    if (document.source === void 0 && opts.resolve?.documentUri !== void 0) {
      (document as Omit<Document, 'source'> & { source: string }).source = opts.resolve?.documentUri;
    }

    const inventory = new DocumentInventory(document, this._resolver);
    await inventory.resolve();

    const validationResults: IRuleResult[] = [...inventory.diagnostics, ...document.diagnostics, ...inventory.errors];

    if (document.formats === void 0) {
      const registeredFormats = Object.keys(this.formats);
      const foundFormats = registeredFormats.filter(format => this.formats[format](inventory.resolved));
      if (foundFormats.length === 0 && opts.ignoreUnknownFormat !== true) {
        document.formats = null;
        if (registeredFormats.length > 0) {
          validationResults.push(this._generateUnrecognizedFormatError(document));
        }
      } else {
        document.formats = foundFormats;
      }
    }

    return {
      resolved: inventory.resolved,
      results: prepareResults(
        [...validationResults, ...runRules(inventory, this.rules, this.functions, this.exceptions)],
        this._computeFingerprint,
      ),
    };
  }

  public async run(target: IParsedResult | Document | object | string, opts: IRunOpts = {}): Promise<IRuleResult[]> {
    return (await this.runWithResolved(target, opts)).results;
  }

  public setFunctions(functions: FunctionCollection) {
    empty(this.functions);

    Object.assign(this.functions, { ...defaultFunctions, ...functions });
  }

  public setRules(rules: RuleCollection) {
    empty(this.rules);

    for (const name in rules) {
      if (!rules.hasOwnProperty(name)) continue;
      const rule = rules[name];

      this.rules[name] = {
        name,
        ...rule,
        severity: rule.severity === void 0 ? DEFAULT_SEVERITY_LEVEL : getDiagnosticSeverity(rule.severity),
      };
    }
  }

  public mergeRules(rules: PartialRuleCollection) {
    for (const name in rules) {
      if (!rules.hasOwnProperty(name)) continue;
      const rule = rules[name];
      if (rule) {
        this.rules[name] = merge(this.rules[name], rule);
      }
    }
  }

  private setExceptions(exceptions: RulesetExceptionCollection) {
    const target: RulesetExceptionCollection = {};
    mergeExceptions(target, exceptions);

    empty(this.exceptions);

    Object.assign(this.exceptions, target);
  }

  public async loadRuleset(uris: string[] | string, options?: IRulesetReadOptions) {
    this.setRuleset(await readRuleset(Array.isArray(uris) ? uris : [uris], options));
  }

  public setRuleset(ruleset: IRuleset) {
    this.setRules(ruleset.rules);

    this.setFunctions(
      Object.entries(ruleset.functions).reduce<FunctionCollection>(
        (fns, [key, { code, ref, name, source, schema }]) => {
          if (code === void 0) {
            if (ref !== void 0) {
              ({ code } = ruleset.functions[ref]);
            }
          }

          if (code === void 0) {
            // shall we log or sth?
            return fns;
          }

          fns[key] = compileExportedFunction(code, name, source, schema);
          return fns;
        },
        {
          ...defaultFunctions,
        },
      ),
    );

    this.setExceptions(ruleset.exceptions);
  }

  public registerFormat(format: string, fn: FormatLookup) {
    this.formats[format] = fn;
  }

  private _generateUnrecognizedFormatError(document: IDocument): IRuleResult {
    return generateDocumentWideResult(
      document,
      `The provided document does not match any of the registered formats [${Object.keys(this.formats).join(', ')}]`,
      DiagnosticSeverity.Warning,
      'unrecognized-format',
    );
  }
}
