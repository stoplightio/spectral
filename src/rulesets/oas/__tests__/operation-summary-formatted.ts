import { DiagnosticSeverity } from '@stoplight/types';
import { RuleType, Spectral } from '../../../spectral';
import * as ruleset from '../index.json';

describe('operation-summary-formatted', () => {
  const s = new Spectral();
  s.addRules({
    'operation-summary-formatted': Object.assign(ruleset.rules['operation-summary-formatted'], {
      recommended: true,
      type: RuleType[ruleset.rules['operation-summary-formatted'].type],
    }),
  });

  test('validate a correct object', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: {
        '/todos': {
          get: {
            summary: 'This is a valid summary.',
          },
        },
      },
    });
    expect(results.length).toEqual(0);
  });

  test('return errors if summary does not start with an uppercase', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: {
        '/todos': {
          get: {
            summary: 'this is not a valid summary.',
          },
        },
      },
    });
    expect(results).toEqual([
      {
        code: 'operation-summary-formatted',
        message: 'Operation `summary` should start with upper case and end with a dot.',
        path: ['paths', '/todos', 'get', 'summary'],
        range: {
          end: {
            character: 49,
            line: 5,
          },
          start: {
            character: 19,
            line: 5,
          },
        },
        severity: DiagnosticSeverity.Warning,
      },
    ]);
  });

  test('return errors if summary does not end with a dot', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: {
        '/todos': {
          get: {
            summary: 'This is not a valid summary',
          },
        },
      },
    });
    expect(results).toEqual([
      {
        code: 'operation-summary-formatted',
        message: 'Operation `summary` should start with upper case and end with a dot.',
        path: ['paths', '/todos', 'get', 'summary'],
        range: {
          end: {
            character: 48,
            line: 5,
          },
          start: {
            character: 19,
            line: 5,
          },
        },
        severity: DiagnosticSeverity.Warning,
      },
    ]);
  });
});
