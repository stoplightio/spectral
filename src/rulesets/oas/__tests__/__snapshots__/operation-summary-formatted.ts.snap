// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operation-summary-formatted return errors if summary does not end with a dot 1`] = `
Array [
  Object {
    "code": "operation-summary-formatted",
    "message": "Operation \`summary\` should start with upper case and end with a dot.",
    "path": Array [
      "paths",
      "/todos",
      "get",
      "summary",
    ],
    "range": Object {
      "end": Object {
        "character": 48,
        "line": 5,
      },
      "start": Object {
        "character": 19,
        "line": 5,
      },
    },
    "severity": 1,
    "source": undefined,
    "summary": "Operation \`summary\` should start with upper case and end with a dot.",
  },
]
`;

exports[`operation-summary-formatted return errors if summary does not start with an uppercase 1`] = `
Array [
  Object {
    "code": "operation-summary-formatted",
    "message": "Operation \`summary\` should start with upper case and end with a dot.",
    "path": Array [
      "paths",
      "/todos",
      "get",
      "summary",
    ],
    "range": Object {
      "end": Object {
        "character": 49,
        "line": 5,
      },
      "start": Object {
        "character": 19,
        "line": 5,
      },
    },
    "severity": 1,
    "source": undefined,
    "summary": "Operation \`summary\` should start with upper case and end with a dot.",
  },
]
`;
